nvflare/private/fed/utils/fed_utils.py

```diff
@@ -446,3 +450,17 @@ def get_job_launcher(job_meta: dict, fl_ctx: FLContext) -> JobLauncherSpec:
         raise RuntimeError(f"The job launcher must be JobLauncherSpec but got {type(launcher)}")
 
     return job_launcher[0]
+
+
+def execute_command_directly(command: str) -> str:
+    """Execute a command directly, without using shell"""
+    args = shlex.split(command)
+
+    if args[0] not in ALLOWED_SHELL_COMMANDS:
+        return f"Error: {args[0]} is not allowed"
```

Comment:
good idea here, should we also add some simple verification for other args for example:

```
def is_valid_arg(arg):
    forbidden = [';', '&', '|', '`', '$', '>', '<']
    return not any(c in arg for c in forbidden)

def execute_command_directly(command: str) -> str:
    args = shlex.split(command)
    cmd = args[0]

    if cmd not in ALLOWED_SHELL_COMMANDS:
        return f"Error: {cmd} is not allowed"

    if not all(is_valid_arg(arg) for arg in args[1:]):
        return "Error: invalid argument detected"
```

and could you add some unit tests for this new method to verify its effectiveness?

Reply:
All those forbidden characters may be valid in grep pattern so this may not work. Adding validation to the code and that may address your concerns.